### ISSUE TITLE
How do I move the keyboard cursor towards the end of a SearchDelegate TextField?

### ISSUE THREAD

**Issue by raymondlee00** [flutter/flutter#51718]
![image](https://user-images.githubusercontent.com/48066594/75617828-8240d200-5b32-11ea-889a-e8a893386f1c.png)

# Code
```
import 'dart:math';

import 'package:Example/models/model.dart';
import 'package:Example/screens/error_screen.dart';
import 'package:Example/util/app_localizations.dart';
import 'package:Example/util/user_util.dart';
import 'package:flutter/material.dart';

class Search<T extends Model<T>> extends SearchDelegate {
  BuildContext context;
  String searchLabel;
  Function view;

  Search(this.context, this.searchLabel, this.view);

  @override
  List<Widget> buildActions(BuildContext context) {
    return [
      IconButton(
        icon: Icon(Icons.clear),
        onPressed: () {
          query = '';
        },
      ),
    ];
  }

  @override
  Widget buildLeading(BuildContext context) {
    return IconButton(
      icon: Icon(Icons.arrow_back),
      onPressed: () {
        close(context, null);
      },
    );
  }

  @override
  Widget buildResults(BuildContext context) {
    if (query.isEmpty)
      showSuggestions(
          context); // TODO: make keyboard unfocus cleaner when searching with empty query

    UserUtil.cacheSearchQuery<T>(query);

    return StatefulBuilder(builder: (
      BuildContext context,
      StateSetter setState,
    ) {
      return StreamBuilder<List<T>>(
        stream: Function.apply(
          view,
          [query],
        ),
        builder: (BuildContext context, AsyncSnapshot<List<T>> snapshot) {
          switch (snapshot.connectionState) {
            case ConnectionState.none:
              return ErrorScreen(
                notifyParent: () => setState(() {}),
                error: snapshot.error,
              );
            case ConnectionState.active:
              return Center(
                child: CircularProgressIndicator(),
              );
            case ConnectionState.waiting:
              return Center(
                child: CircularProgressIndicator(),
              );
            case ConnectionState.done:
              if (snapshot.hasError) {
                return ErrorScreen(
                  notifyParent: () => setState(() {}),
                  error: snapshot.error,
                );
              }
              return ListView(
                children: snapshot.data.map((obj) => obj.toCard()).toList(),
              );
          }
        },
      );
    });
  }

  @override
  Widget buildSuggestions(BuildContext context) {
    return StatefulBuilder(
      builder: (BuildContext context, StateSetter setState) {
        return FutureBuilder(
          future: UserUtil.getCachedSearches<T>(),
          builder: (context, snapshot) {
            switch (snapshot.connectionState) {
              case ConnectionState.waiting:
                return Container();
              default:
                if (snapshot.hasError) {
                  return ErrorScreen(
                    notifyParent: () => setState(() {}),
                    error: snapshot.error,
                  );
                }
                return ListView.builder(
                  itemCount: snapshot.data.length,
                  itemBuilder: (context, index) {
                    return ListTile(
                      leading: IconButton(
                        icon: Icon(Icons.search),
                        onPressed: () => {
                          query = snapshot.data[index],
                          showResults(context),
                        },
                      ),
                      trailing: Transform.rotate(
                        angle: 270 * pi / 180,
                        child: IconButton(
                          icon: Icon(
                            Icons.call_made,
                          ),
                          onPressed: () => {
                            query = snapshot.data[
                                index], // TODO: shows cursor in the beginning of query, which looks weird
                          },
                        ),
                      ),
                      title: Text('${snapshot.data[index]}'),
                      onTap: () => {
                        query = snapshot.data[index],
                        showResults(context),
                      },
                    );
                  },
                );
            }
          },
        );
      },
    );
  }

  @override
  String get searchFieldLabel =>
      AppLocalizations.of(context).translate(searchLabel);
}
```

## Flutter Doctor
```
[√] Flutter (Channel stable, v1.12.13+hotfix.8, on Microsoft Windows [Version 10.0.18363.657], locale en-US)
    • Flutter version 1.12.13+hotfix.8 at C:\Users\OWNER\flutter
    • Framework revision 0b8abb4724 (3 weeks ago), 2020-02-11 11:44:36 -0800
    • Engine revision e1e6ced81d
    • Dart version 2.7.0

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\OWNER\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.42.1)
    • VS Code at C:\Users\OWNER\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.8.1

[√] Connected device (1 available)
    • Alcatel 4060A • 582caa84 • android-arm • Android 5.1.1 (API 22)

• No issues found!
```

I looked at the source search.dart file and it seems that the TextEditingController `final TextEditingController _queryTextController = TextEditingController();
` is private and cannot be changed from outside the file.
Since modifying the source code would seem too troublesome, I am considering copying the source search.dart file and modifying the private variables to control the cursor. Is there any way other than tampering with the source code?

**Comment by TahaTesser** [comment]
Hi @raymondlee00 
Please try this possible  [solution](https://github.com/flutter/flutter/issues/11416#issuecomment-373635441)
Thanks

**Comment by raymondlee00** [comment]
> Hi @raymondlee00
> Please try this possible [solution](https://github.com/flutter/flutter/issues/11416#issuecomment-373635441)
> Thanks

I have found this solution but how would I get the _textController inside of the SearchDelegate class? The entire solution depends on being able to access the _textController in the first place.

**Comment by darshankawar** [comment]
Hi @raymondlee00,
Can you please confirm if you still experience this issue on latest Flutter version (eg: dev, master) ? Seeing your last message, it seems you found the solution. Is your other question still open ?

**Comment by raymondlee00** [comment]
I never found the solution. I haven't been working on the project since mid-March so I haven't tried to fix it since then. I don't know if I would still experience this issue on the latest Flutter version.

**Comment by darshankawar** [comment]
Hi @raymondlee00,
Can you make some time and try to confirm on latest Flutter version to see if error is gone or still exists ?

**Comment by markusaksli-nc** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

**Comment by mrcnserkan** [comment]
You can call this function by adding it to the main SearchDelegate class (flutter/packages/flutter/lib/src/material/search.dart) :
`void setCursorPosition(int index) {
    _queryTextController.selection = TextSelection.fromPosition(TextPosition(offset: index));
  }`

**Comment by mrcnserkan** [comment]
> I never found the solution. I haven't been working on the project since mid-March so I haven't tried to fix it since then. I don't know if I would still experience this issue on the latest Flutter version.

You can call this function by adding it to the main SearchDelegate class (flutter/packages/flutter/lib/src/material/search.dart) :
`void setCursorPosition(int index) {`
   ` _queryTextController.selection = TextSelection.fromPosition(TextPosition(offset: index));`
 ` }`

**Comment by alecerri** [comment]
I see this problem still exists on ver 1.22.1 of flutter and unfortunately the root cause is that _queryTextController is still private hence it is impossible to control the cursor. Any chance you reopen this ticket and add the possibility to set the cursor?

**Comment by TahaTesser** [comment]
Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.

**Comment by intraector** [comment]
Same here

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.4, on Mac OS X 10.14.6 18G103 darwin-x64, locale en-RU)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[!] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to
        your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install:
        sudo gem install cocoapods
[!] Android Studio (version 4.1)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.51.1)
[✓] Connected device (1 available)
```

**Comment by ajmalsalim** [comment]
This still exists.

The workaround was: https://github.com/flutter/flutter/issues/11416#issuecomment-405792611 but since the TextController is not exposed, this is not possible without modifying the search.dart file.

The exact issue is when we select a suggestion item from buildSuggestions() and if the query is empty, we try to set the query using query= 'suggestion 1'; This will set the value in the search box but the cursor will be at the beginning. 

In other words, the set query(String value) method is setting the text but the cursor is at the beginning.


`✓] Flutter (Channel stable, 1.22.4, on Mac OS X 10.15.7 19H2 darwin-x64, locale en-AU)
    • Flutter version 1.22.4 at /Users/user/Flutter/flutter
    • Framework revision 1aafb3a8b9 (3 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4`

**Comment by garan82** [comment]
I also need to be able to access to _queryTextController in SearchDelegate, which is currently not possible. It's for another use case. Please triage this.

**Comment by Nuke-101** [comment]
> 
> 
> > I never found the solution. I haven't been working on the project since mid-March so I haven't tried to fix it since then. I don't know if I would still experience this issue on the latest Flutter version.
> 
> You can call this function by adding it to the main SearchDelegate class (flutter/packages/flutter/lib/src/material/search.dart) :
> `void setCursorPosition(int index) {`
> ` _queryTextController.selection = TextSelection.fromPosition(TextPosition(offset: index));`
> ` }`

This solved my issue thanks :)

**Comment by DarkMikey** [comment]
Please reopen & make SearchDelegate's `_queryTextController` public or find some other way to solve that behaviour.

**Comment by TahaTesser** [comment]
@DarkMikey 
For any proposal, please file a new issue as proposal

**Comment by bikcrum** [comment]
This has been resolved by #83512
